File: src/waldiez/models/Tool/Tool.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2024 - 2025 Waldiez & contributors
 */
import { WaldiezToolData } from "@waldiez/models/Tool/ToolData";
import { getId } from "@waldiez/utils";

/**
 * Waldiez Tool
 * @param type - The type (tool)
 * @param id - The ID
 * @param name - The name of the tool
 * @param description - The description of the tool
 * @param tags - The tags
 * @param requirements - The requirements
 * @param createdAt - The created at date
 * @param updatedAt - The updated at date
 * @param data - The data
 * @param rest - Any additional properties
 * @see {@link WaldiezToolData}
 */
export class WaldiezTool {
    type = "tool";
    id: string;
    name: string;
    description: string;
    tags: string[];
    requirements: string[];
    createdAt: string;
    updatedAt: string;
    data: WaldiezToolData;
    rest?: { [key: string]: unknown } = {};

    constructor(props: {
        id: string;
        data: WaldiezToolData;
        name: string;
        description: string;
        tags: string[];
        requirements: string[];
        createdAt: string;
        updatedAt: string;
        rest?: { [key: string]: unknown };
    }) {
        this.id = props.id;
        this.name = props.name;
        this.description = props.description;
        this.tags = props.tags;
        this.requirements = props.requirements;
        this.createdAt = props.createdAt;
        this.updatedAt = props.updatedAt;
        this.data = props.data;
        this.rest = props.rest;
    }

    static create(): WaldiezTool {
        return new WaldiezTool({
            id: `wt-${getId()}`,
            name: "new_tool",
            description: "A new tool",
            tags: [],
            requirements: [],
            createdAt: new Date().toISOString(),
            updatedAt: new Date().toISOString(),
            data: new WaldiezToolData(),
        });
    }
}

File: tests/exporting/flow_helpers/tool.py
# SPDX-License-Identifier: Apache-2.0.
# Copyright (c) 2024 - 2026 Waldiez and contributors.
# Copyright (c) 2024 - 2025 Waldiez and contributors.
# pyright: reportCallIssue=false
"""Helper functions for Waldiez models."""

from typing_extensions import Literal

from waldiez.models import (
    WaldiezTool,
    WaldiezToolData,
    WaldiezToolType,
)


def get_tool(
    tool_id: str = "wt-1",
    tool_type: WaldiezToolType = "custom",
) -> WaldiezTool:
    """Get a WaldiezTool.

    Parameters
    ----------
    tool_id : str, optional
        The tool ID, by default "wt-1"
    tool_type : WaldiezToolType, optional
        The tool type, by default "custom"

    Returns
    -------
    WaldiezTool
        A WaldiezTool instance.
    """
    return WaldiezTool(
        id=tool_id,
        name="tool_name",
        description="Tool Description",
        tags=["tool"],
        requirements=["chess"],
        type="tool",
        created_at="2021-01-01T00:00:00.000Z",
        updated_at="2021-01-01T00:00:00.000Z",
        data=WaldiezToolData(
            tool_type=tool_type,
            content=(
                "def tool_name() -> str:\n"
                '    """Tool Description."""\n'
                "    return 'Tool Response'"
            ),
            secrets={
                "TOOL_KEY": "tool_value",
            },
        ),
    )


def get_interop_tool(
    tool_id: str = "wt-2",
    tool_type: Literal["langchain", "crewai"] = "langchain",
) -> WaldiezTool:
    """Get an interop tool.

    Parameters
    ----------
    tool_id : str, optional
        The tool ID, by default "wt-2"
    tool_type : Literal["langchain", "crewai"]
        The tool type, by default "langchain"

    Returns
    -------
    WaldiezTool
        A WaldiezTool instance.
    """
    tool_name = f"{tool_type}_tool"
    # noinspection PyTypeChecker
    return WaldiezTool(
        id=tool_id,
        name=tool_name,
        description="Interop Tool Description",
        tags=["interop_tool"],
        requirements=[],
        type="tool",
        created_at="2021-01-01T00:00:00.000Z",
        updated_at="2021-01-01T00:00:00.000Z",
        data=WaldiezToolData(
            content=f"{tool_name} = lambda: 'Interop Tool Response'",
            tool_type=tool_type,
            secrets={},
        ),
    )
